feat: [TD-277] create Note model

ross-garcia opened the pull request:
## Summary

Archivos añadidos:
- modelo Note
- migración para Note
- spec/factories/notes.rb
- Test para el modelo Note

Archivos modificados: 
- modelo User para agregar la relación con Note

## Screenshots

Tests:

![image](https://github.com/ross-garcia/Training-Rails-UGUITO-API-Bootstrap/assets/160248891/c068e6c4-b7d2-4b6f-bc9c-ae7ad6bdaf71)

Pruebas por consola:

![image](https://github.com/ross-garcia/Training-Rails-UGUITO-API-Bootstrap/assets/160248891/7b9dc2c2-d251-4d21-9064-d9f28f78c133)

Se omite el campo de utility_id porque la realación con Utility es por medio de User haciendo uso de has_one through

## Test Cases
N/A

## Checklist 

- [+ ] Revisé los archivos modificados antes de liberar el pull request para revisión. 👩🏻‍💻
- [+ ] Actualicé la card de JIRA considerando: **estado**, **horas incurridas**, **enfoque tomado como comentario de la card** y **requisitos de configuración asociados**✅
- [+ ] El título de mi PR sigue la convención requerida 👉 [%scope - %descripcion](https://www.conventionalcommits.org/en/v1.0.0/)

## JIRA Card

https://widergy.atlassian.net/browse/**[TD-277]**

ross-garcia committed:
feat: create Note model
- Note migration added
- Note factory spec added
- Test for Note model added
- User model modified (has_many association with Note)

ross-garcia committed:
feat: Merge branch 'master' into TD-277-note-model

JoaquinSeita commented on db/migrate/20240220125322_create_notes.rb:
```diff
@@ -0,0 +1,12 @@
+class CreateNotes < ActiveRecord::Migration[6.1]
+  def change
+    create_table :notes do |t|
+      t.belongs_to :user, foreign_key: true
```
Todas las notas deben pertenecer a un usuario, fijate que lo estas validando en el modelo, sin embargo en la migración y en el `schema.rb` podría ser nulo

ross-garcia replied:
Hola @JoaquinSeita  modifiqué la migración de Note para evitar que user_id sea nulo, también eliminé el valor por defecto.

JoaquinSeita commented on db/migrate/20240220125322_create_notes.rb:
```diff
@@ -0,0 +1,12 @@
+      t.belongs_to :user, foreign_key: true
+      t.string :title, null: false
+      t.string :content, null: false
+      t.integer :note_type, default: 0, null: false
```
Entiendo la inciativa! Pero con las siguientes cards ese `default` no sera necesario

JoaquinSeita commented on db/schema.rb:
```diff
@@ -75,6 +75,16 @@
   end
 
+  create_table "notes", force: :cascade do |t|
+    t.bigint "user_id"
```
Acá se ve lo del `user_id` que te comente antes

JoaquinSeita commented on spec/factories/notes.rb:
```diff
@@ -0,0 +1,8 @@
+    user
+    title { Faker::Book.title }
+    content { Faker::Lorem.sentence }
+    note_type { :review }
```
`note_type` es de tipo `enum`, podes acceder a los valores del mismo si le das una miradita a esa doc:
https://blog.saeloun.com/2022/01/05/how-to-use-enums-in-rails/

Y combinando con estos métodos, podrías crear notas con los tipos de forma aleatoria:
https://apidock.com/ruby/Hash/values y https://apidock.com/ruby/Array/sample

ross-garcia replied:
Ya hice el refactoring de note_type, me ayudó los enlaces, gracias:)

JoaquinSeita commented on spec/models/note_spec.rb:
```diff
@@ -0,0 +1,19 @@
+require 'rails_helper'
+
+RSpec.describe Note, type: :model do
+  subject(:note) do
+    create(:note)
+  end
```
Para este caso con hacer un `build(:note)` alcanza y tambien podes usar la sintaxis de llaves para ahorrarte unas lineas:
```suggestion
  subject(:note) { build(:note) }
```

ross-garcia replied:
@JoaquinSeita intenté probar con build y me fallaban los test debido a la relación que tiene con User. Por esta razón mantuve el "create". Lo revisé con Flor y me explicó que se debe a que build realmente no está creando el registro en la DB y por eso no genera la asociación con User.

JoaquinSeita commented on spec/models/note_spec.rb:
```diff
@@ -0,0 +1,19 @@
+    it { is_expected.to validate_presence_of(value) }
+  end
+
+  it { expect(subject).to define_enum_for(:note_type).with_values({ review: 0, critique: 1 }) }
+
+  it { is_expected.to belong_to(:user) }
```
Podrías agregar un test comprobar que el modelo tenga asociada una `Utility` a través de `User`

JoaquinSeita reviewed (commented):
Buen laburo Ross! :100: Te dejo algunos comentarios, cualquier cosa mandame un mensaje  si necesitas una mano

ross-garcia committed:
feat: code corrections:
- Note migrations modified
- Note model modified
- test for Note model added
- refactoring Note factory

JoaquinSeita reviewed (approved):
Genial! :100: Muchas gracias por responder los comentarios, se ve bien para mi :smiley:
